lib/phoenix_live_view.ex

```diff
@@ -474,51 +474,64 @@ defmodule Phoenix.LiveView do
       @before_compile Phoenix.LiveView.Renderer
 
       @phoenix_live_opts opts
-      Module.register_attribute(__MODULE__, :mount, accumulate: true)
+      Module.register_attribute(__MODULE__, :phoenix_live_on_mount, accumulate: true)
       @before_compile Phoenix.LiveView
     end
   end
 
-  defmacro __before_compile__(env) do
-    opts = Module.get_attribute(env.module, :phoenix_live_opts, [])
-    at_mount = Module.get_attribute(env.module, :mount, [])
-
-    quote bind_quoted: [opts: opts, at_mount: at_mount] do
+  defmacro __before_compile__(_env) do
+    quote do
       @doc false
       def __live__ do
-        unquote(Macro.escape(Phoenix.LiveView.__live__(__MODULE__, at_mount, opts)))
-      end
-    end
-  end
+        alias Phoenix.LiveView.Lifecycle
+
+        module = __MODULE__
+        opts = @phoenix_live_opts
+
+        container = opts[:container] || {:div, []}
+
+        namespace =
+          opts[:namespace] || module |> Module.split() |> Enum.take(1) |> Module.concat()
+
+        name = module |> Atom.to_string() |> String.replace_prefix("#{namespace}.", "")
 
-  @doc false
-  def __live__(module, at_mount, opts) do
-    container = opts[:container] || {:div, []}
-    namespace = opts[:namespace] || module |> Module.split() |> Enum.take(1) |> Module.concat()
-    name = module |> Atom.to_string() |> String.replace_prefix("#{namespace}.", "")
+        layout =
+          case opts[:layout] do
+            {mod, template} when is_atom(mod) and is_binary(template) ->
+              {mod, template}
 
-    layout =
-      case opts[:layout] do
-        {mod, template} when is_atom(mod) and is_binary(template) ->
-          {mod, template}
+            nil ->
+              nil
 
-        nil ->
-          nil
+            other ->
+              raise ArgumentError,
+                    ":layout expects a tuple of the form {MyLayoutView, \"my_template.html\"}, " <>
+                      "got: #{inspect(other)}"
+          end
+
+        lifecycle = Lifecycle.mount(module, @phoenix_live_on_mount)
 
-        other ->
-          raise ArgumentError,
-                ":layout expects a tuple of the form {MyLayoutView, \"my_template.html\"}, " <>
-                  "got: #{inspect(other)}"
+        %{
+          container: container,
+          name: name,
+          kind: :view,
+          module: module,
+          layout: layout,
+          lifecycle: lifecycle
+        }
       end
+    end
+  end
```

Comment:
Sorry, I gave confusing advice. We don't want to compute this inside `__live__`, just rather inside the __before_compile__ body. Something like this:

```elixir
  defmacro __before_compile__(env) do
    opts = Module.get_attribute(env.module, :phoenix_live_opts)

    layout =
      case opts[:layout] do
        {mod, template} when is_atom(mod) and is_binary(template) ->
          {mod, template}

        nil ->
          nil

        other ->
          raise ArgumentError,
                ":layout expects a tuple of the form {MyLayoutView, \"my_template.html\"}, " <>
                  "got: #{inspect(other)}"
      end

    phoenix_live_on_mount = Module.get_attribute(env.module, :phoenix_live_on_mount)
    lifecycle = Phoenix.LiveView.Lifecycle.mount(module, phoenix_live_on_mount)

    container = opts[:container] || {:div, []}
    name = module |> Atom.to_string() |> String.replace_prefix("#{namespace}.", "")

    namespace =
      opts[:namespace] || module |> Module.split() |> Enum.take(1) |> Module.concat()

    live = %{
      container: container,
      name: name,
      kind: :view,
      module: module,
      layout: layout,
      lifecycle: lifecycle
    }

    quote do
      @doc false
      def __live__, do: unquote(Macro.escape(live))
    end
  end
```

I haven't tested it though!

Reply:
Ah yes, makes much more sense this way :)